### PR TITLE
Fix: scope ProcessedMail UID deduplication by IMAP UIDVALIDITY

### LINKED ISSUE #8939: emails are marked as 'Already Processed' if the imap server reuses the same IMAP id after deletion.
### Description

I was trying to set up a micro mailserver using GNU mailutils imap4d server but only the first few mails were processed and deleted, eventually I found out that imap4d uses the X-UID header to store the id and always assigns lowest free number starting from 1, so if the INBOX is empty an email alway gets 1 as IMAP UID.
If I edit manually the mbox file on the server and assign a different X-UID the email is correctly processed by paperless.
Other imap server, like dovecot, use more complex uid and it is less likely that they are reused but there is no guarantee that it doesn't happen. Also I empty my mailbox once a year so even if paperless doesn't delete imediately processed emails there is no guarantee that the UID is not eventually re-used in future if the user deletes the email.
UID is the unique identification number of a email in a IMAP folder, UID's are assigned by a imap server but there is no guarantee that they are not re-used. It is not a reliable way to determine if an email was altready processed expecially if paperless is configured to delete the email once it is processed.

### Steps to reproduce

Send email with scans to imap4d server let paperless process and delete it
 [DEBUG] [paperless_mail] Rule paperless consume.paperless consume: Processing mail scansioni nuove from xxxx@xxxx.xxx.xx with 2 attachment(s)
send another email with scans to the same imap4d server and try to let paperless read it:
 [DEBUG] [paperless_mail] Skipping mail '1' subject 'scansioni nuovissime' from 'xxxx@xxxx.xxx.xx', already processed.
edit the mbox changing the X-UID header to a different number, restart the imap server let paperless scan again the mailbox
[DEBUG] [paperless_mail] Rule paperless consume.paperless consume: Processing mail scansioni nuovissime from xxxx@xxxx.xxx.xx with 2 attachment(s)
this time the email is correctly processed.

### Webserver logs

```bash
2025-01-29 14:40:03,641] [DEBUG] [paperless_mail] Processing mail account paperless consume

[2025-01-29 14:40:03,750] [DEBUG] [paperless_mail] GMAIL Label Support: False

[2025-01-29 14:40:03,751] [DEBUG] [paperless_mail] AUTH=PLAIN Support: True

[2025-01-29 14:40:08,121] [DEBUG] [paperless_mail] Account paperless consume: Processing 1 rule(s)

[2025-01-29 14:40:08,126] [DEBUG] [paperless_mail] Rule paperless consume.paperless consume: Selecting folder INBOX

[2025-01-29 14:40:12,041] [DEBUG] [paperless_mail] Rule paperless consume.paperless consume: Searching folder with criteria (SINCE 30-Dec-2024)

[2025-01-29 14:40:14,048] [DEBUG] [paperless_mail] Skipping mail '1' subject 'scansioni nuove' from 'xxxx@xxxx.xxx.xx', already processed.

[2025-01-29 14:40:14,050] [DEBUG] [paperless_mail] Rule paperless consume.paperless consume: Processed 0 matching mail(s)
```

### Browser logs

```bash

```

### Paperless-ngx version

2.14.6

### Host OS

docker

### Installation method

Docker - official image

### System status

```json

```

### Browser

chrome

### Configuration changes

_No response_

### Please confirm the following

- [x] I believe this issue is a bug that affects all users of Paperless-ngx, not something specific to my installation.
- [x] This issue is not about the OCR or archive creation of a specific file(s). Otherwise, please see above regarding OCR tools.
- [x] I have already searched for relevant existing issues and discussions before opening this report.
- [x] I have updated the title field above with a concise description.

### PR DESCRIPTION
**Pull request by chouquette**
## Proposed change

Paperless dedupes incoming mail by `(rule, folder, uid)`, but IMAP UIDs are only unique within a UIDVALIDITY epoch. Any server that rotates UIDVALIDITY while reassigning sequential UIDs from a low number — ProtonMail Bridge on restart, or GNU mailutils `imap4d` after manual mailbox manipulation (#8939) — will have new mail silently logged as "already processed" and dropped.

This was previously reported in #8939 and closed as an IMAP server-side violation of RFC 3501 §2.3.1.1. Reading the relevant section again, I think the responsibility actually lies with the client: servers are permitted to rotate UIDVALIDITY provided the new value is greater than the previous one, and the protocol asks clients to invalidate cached UIDs when that happens.

> If unique identifiers from an earlier session fail to persist in this session, the unique identifier validity value **MUST be greater** than the one used in the earlier session.
> — [RFC 3501 §2.3.1.1](https://www.rfc-editor.org/rfc/rfc3501.html#section-2.3.1.1)

### Fix

Scope `ProcessedMail` deduplication by UIDVALIDITY in addition to UID. The new field is nullable and existing rows are left untouched, so no data migration is needed. 
The two Celery shared tasks gain an empty default for the new parameter so queued tasks from the previous version don't crash on upgrade.

Should fix #8939.

### Testing

- Two regression tests for the dedupe path: skip when UIDVALIDITY is unchanged, process when it has changed.
- I send scans directly from my scanner to a dedicated email account that's set up in paperless-ngx. Previously all mails were discarded as already imported. With this patch, they are imported as expected.

### AI disclosure

The code was fully authored by Claude Code. I reviewed it and directed the instructions so that the fix is as non-intrusive as possible, and tested that it fixes my usecase, but it's still 100% AI generated.

I'm not sure what's meant by "PRs that are entirely or mostly AI-derived", but this PR was created manually, to solve a specific problem on my (human) setup.

## Type of change

- [x] Bug fix: non-breaking change which fixes an issue.
- [ ] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [ ] If applicable, I have tested my code for breaking changes & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all Git `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] In the description of the PR above I have disclosed the use of AI tools in the coding of this PR.

### CODE REVIEW

**Comment by shamoon** [comment]
> The code was fully authored by Claude Code.

> I'm not sure what's meant by "PRs that are entirely or mostly AI-derived

Ha, this.

Thanks, we’ll look when we have a sec.

**Comment by chouquette** [comment]
> I'm not sure what's meant by "PRs that are entirely or mostly AI-derived

To clarify what I meant, I don't know if this is meant to target fully automated PRs based on scanners or whatnot.

I just wanted to clarify that this is not a PR generated casting a large net and hoping something will be generated, in order to bump my stats or similar. It fixes an issue I'm seeing in my usage of Paperless-NGX which seems to be upstreamable IMHO

**Review by codeCraft-Ritik — COMMENTED** [review]
Solid bug fix — this addresses a subtle but important edge case around IMAP UID reuse.
The regression tests and minimal migration impact make the approach feel low-risk.

File: src/paperless_mail/mail.py [hunk 1]
```diff
@@ -646,6 +655,10 @@ def _handle_mail_rule(
                 f"does not exist in account {rule.account}",
             ) from err
 
+        self._current_uid_validity = str(
+            M.folder.status(rule.folder, ["UIDVALIDITY"])["UIDVALIDITY"],
```
Comment (stumpylog) [inline src/paperless_mail/mail.py]: If this fails, the entire message is not consumed.  Either due to a network error, status not supported or `UIDVALIDITY` not provided.  That basically disables the entire rule, which we do not want.  Errors should be logged, handled and the UIDVALIDITY assume to be None or empty or some other sentinel value.

File: src/paperless_mail/models.py [hunk 2]
```diff
@@ -338,6 +338,14 @@ class ProcessedMail(document_models.ModelWithOwner):
         editable=False,
     )
 
+    uid_validity = models.CharField(
```
Comment (stumpylog) [inline src/paperless_mail/models.py:341]: This does not work.  All existing processed mail rows will have a NULL value here, which means `ProcessedMail.objects.filter` is not going to see those, resulting in processing of basically all already processed emails.

File: src/paperless_mail/mail.py
Comment (stumpylog) [inline src/paperless_mail/mail.py]: Changing the default to None, then updating this to

```python
  ProcessedMail.objects.filter(
      rule=rule, uid=message.uid, folder=rule.folder,
  ).filter(
      Q(uid_validity=current_uid_validity) | Q(uid_validity__isnull=True)
  ).exists()
```

should prevent re-ingestion.

File: src/paperless_mail/mail.py [hunk 3]
```diff
@@ -313,6 +316,7 @@ def error_callback(
     message_uid: str,
     message_subject: str,
     message_date: datetime.datetime,
+    uid_validity: str = "",
```
Comment (stumpylog) [inline src/paperless_mail/mail.py]: Make this `str | None = None`

File: src/paperless_mail/mail.py [hunk 4]
```diff
@@ -244,6 +244,7 @@ def apply_mail_action(
...
```
Comment (stumpylog) [inline src/paperless_mail/mail.py]: Make this `str | None = None`

**Review by stumpylog — CHANGES_REQUESTED** [review]

**Comment by stumpylog** [comment]
I'd like to see a test verifying the operation is correct when a prior ProcessedMail row with a real uid_validity exists and the current UIDVALIDITY is unavailable (e.g. the server returns MailboxFolderStatusError, so _get_uid_validity returns None).

The expectation would be that the mail is skipped as already-processed (_queue_consumption_tasks_mock.call_count == 0) - i.e. when the current validity can't be determined, dedup should fall back to the existing (rule, uid, folder) match rather than re-ingesting.  Re-ingesting is something we really, really do not want.

**Review by stumpylog — APPROVED** [review]
